[![Tests]( https://github.com/ivargr/hic-assembly-benchmarking/actions/workflows/test_snakemake_pipeline.yml/badge.svg)](https://github.com/ivargr/hic-assembly-benchmarking/actions/workflows/test_snakemake_pipeline.yml)


## Snakemake pipeline for testing HiC phasing and scaffolding on simulated data

Overview:

* This pipeline can simulate a diploid individual with snps, indels and optinally structural variation on a reference genome of choice. Specify the individual in `config/config.yaml`. 
* Hifi + HiC reads can be simulated from this genome
* Hifiasm is run using the simulated reads
* Yahs and other tools can be run to scaffold the assembly graph created by Hifiasm
* The resulting assembly is evaluated by Quast


### Installation

1. Clone this repo. 
2. If you want to also simulate HiFi reads (not necessary to test simple scaffolding) you unfortunately need to manually install pbsim3. This can be done by running `./install_pbsim.sh`.
2. Install Python requirements: `pip install -r requirements.txt`
3. Test that the pipeline works: `snakemake --use-conda -F plots/test.png` (should finish without errors in ~5 minutes). The plot `plots/test.png` will show scaffolding accuracy for a few simple test cases.


### Make a test plot
```snakemake
snakemake --use-conda plots/test.png
```


### Run with real yeast HiC data
It's possible to run quick tests with real yeast HiC data. See the config in `config/plots.yaml` where plots can be configured. A plot can be generated by running `snakemake plots/PLOT_NAME.png`, e.g.:

```snakemake
snakemake plots/yeast_real_data.png
```

### Snakemake folder structure
The current folder structure is as follows:
`data/genome_build/individual_id/dataset_size/hifi_reads_depth/n_hic_reads/assembled_or_not_assembled/n_artificial_contigs/...`.

For instance, the following runs yahs and runs Edison on the yahs scaffold to produce the `scaffolds.fa` file (the final scaffolds). This is run with simulated hic reads, on the small dataset (specified in config/config.yaml). The contigs are not from an assembly, but simulated with 10 splits (the last 10 in the path)..

```snakemake
snakemake --use-conda data/sacCer3/simulated/small/10/40000/not_assembled/10/yahs/scaffolds.fa
```

The `scaffolds.fa` above can be changed to:

* `report.pdf`: Generates a quast report
* `edison.txt`: Runs Edison and creates a report with accuracy, etc.
* `heatmap.png`: Creates a simple heatmap after mapping the HiC reads to the final scaffolds. 
* `alignment.pdf`: Visualization of how the final scaffolds aligns to the reference 


### Missing functionality / todo
* Genome simulation is hardcoded in config and the same for all simulations. Should be possible to configure for a given run
* Heterozygosity is fixed, should be tunable. Every variant is now default heterozygous. 